File: rss_reader/sync.py
from defusedxml import ElementTree as Xml
import requests
from rss_reader.db import connect
from rss_reader.date import parse_date
import datetime as dt
import os


class Sync:
    def __init__(self):
        self.db = connect()

    def run(self):
        query = "SELECT * FROM blogs LEFT JOIN feeds ON blogs.id == feeds.blog_id"
        for row in self.db.execute(query):
            self.sync(row)

    def read_feed(self, blog: dict):
        response = requests.get(blog["xml_url"])
        return response.text

    def sync(self, blog):
        content = self.read_feed(blog)
        xml = Xml.fromstring(content)

        # for atom feeds
        for post in xml.iter("entry"):
            self.sync_post(blog, post)

        # for rss feeds
        for post in xml.iter("item"):
            self.sync_post(blog, post)

    def sync_post(self, blog, tag):
        """
        Syncs a post from an rss or atom feed.
        """
        post = {"blog_id": blog["id"], "published_at": None}
        for child in tag.iter():
            if child.tag == "title":
                post["title"] = child.text

            elif child.tag == "link":
                post["url"] = child.text

            # rss-only feeds
            elif child.tag == "pubDate":
                post["published_at"] = parse_date(child.text)

            # atom-only feeds
            elif child.tag == "published":
                post["published_at"] = parse_date(child.text)

        try:
            self.db.execute(
                """
                INSERT INTO posts(blog_id, title, url, published_at) 
                VALUES(:blog_id, :title, :url, :published_at)
                ON CONFLICT DO UPDATE SET
                    title=excluded.title,
                    published_at=excluded.published_at
            """,
                post,
            )
        except Exception as e:
            print(f"skipping post: {post}\n{e}")

File: server/sync.ts
import { QueryFunc } from "@/server/define.ts";
import { md5 } from "@takker/md5";
import { encodeHex } from "jsr:@std/encoding@1/hex";
import { parseXmlStream, type XmlEventCallbacks } from "@std/xml";

type Feed = {
  xml_url: string;
  etag?: string;
  last_modified?: string;
  hash?: string;
};

export async function sync(sql: QueryFunc, blogId: number): Promise<void> {
  const feed: Feed = (
    await sql`
    select b.xml_url, f.etag, f.last_modified, f.hash
    from blogs b 
    left join feeds f on b.id = f.blog_id
    where b.id = ${blogId}
  `
  )[0] as unknown as Feed;
  console.log(feed);
  if (feed.xml_url == undefined) {
    console.error("no blog found!");
    return;
  }

  const headers: HeadersInit = {};
  if (feed.etag) {
    headers["If-None-Match"] = feed.etag;
  }

  if (feed.last_modified) {
    headers["If-Modified-Since"] = feed.last_modified;
  }
  const response = await fetch(feed.xml_url, {
    method: "GET",
    headers,
  });

  const body = await response.text();
  // no updates on a not modified
  if (response.status === 304) {
    return;
  }

  const hash = encodeHex(md5(body));
  const shouldUpdatePosts = feed.hash != hash;
  feed.hash = hash;
  feed.etag = response.headers.get("etag")!;
  feed.last_modified = response.headers.get("last-modified")!;
  await sql`
    insert into feeds(blog_id, hash, etag, last_modified)
    values (${blogId}, ${feed.hash}, ${feed.etag}, ${feed.last_modified})
    on conflict(blog_id) do update set
      hash = excluded.hash,
      etag = excluded.etag,
      last_modified = excluded.last_modified
  `;

  if (shouldUpdatePosts) {
    await updatePosts(sql, blogId, body);
  }
}

type Post = {
  title: string;
  url: string;
  published_at_text: string;
  thumbnail: string;
};

export async function force_sync(
  sql: QueryFunc,
  blogId: number,
): Promise<void> {
  const feed = (
    await sql`
    select b.xml_url
    from blogs b 
    where b.id = ${blogId}
  `
  )[0].xml_url as string;
  const response = await fetch(feed);
  const body = await response.text();
  await updatePosts(sql, blogId, body);
}

async function updatePosts(sql: QueryFunc, blogId: number, body: string) {
  const posts: Array<Post> = [];
  let post: Partial<Post> = {};
  let el: string = "";

  const xmlCallbacks: XmlEventCallbacks = {
    onEndElement(name) {
      if (name == "entry" || name == "item") {
        posts.push(post as Post);
      }
    },

    onStartElement(name, _, __, attributes) {
      el = name;
      if (name == "entry" || name == "item") {
        post = {};
      } else if (name == "link") {
        for (let i = 0; i < attributes.count; i++) {
          if (attributes.getName(i) == "href") {
            post.url = attributes.getValue(i);
          }
        }
      } else if (name == "thumbnail") {
        for (let i = 0; i < attributes.count; i++) {
          if (attributes.getName(i) == "url") {
            post.thumbnail = attributes.getValue(i).trim();
          }
        }
      }
    },

    onText(text) {
      const trimmed = text.trim();
      if (trimmed === "") return;

      if (el == "title") {
        post.title = trimmed;
      } else if (el == "link") {
        post.url = trimmed;
      } else if (el == "published" || el == "pubDate") {
        post.published_at_text = trimmed;
      }
    },
  };

  await parseXmlStream(intoStream(body), xmlCallbacks);

  for (const post of posts) {
    if (post.title == null) {
      console.error(post);
      return;
    }
    try {
      await sql`
          insert into posts(blog_id, title, url, published_at_text, thumbnail)
          values (${blogId}, ${post.title}, ${post.url}, ${post.published_at_text ?? null}, ${post.thumbnail ?? null})
          on conflict (url) do update
          set 
            title=excluded.title,
            published_at_text=excluded.published_at_text,
            thumbnail=excluded.thumbnail
        `;
    } catch (e) {
      console.error(e);
      console.log(post);
    }
  }
}

async function* intoStream(body: string) {
  yield body;
}
